core/src/main/java/com/amazon/opendistroforelasticsearch/sql/analysis/ExpressionAnalyzer.java

```diff
@@ -163,6 +169,43 @@ public Expression visitCompare(Compare node, AnalysisContext context) {
         repository.compile(functionName, Arrays.asList(left, right));
   }
 
+  @Override
+  public Expression visitCase(Case node, AnalysisContext context) {
+    List<WhenClause> whens = new ArrayList<>();
+    for (When when : node.getWhenClauses()) {
+      if (node.getCaseValue() == null) {
+        whens.add((WhenClause) analyze(when, context));
+      } else {
+        // Merge case value and condition (compare value) into a single equal condition
+        whens.add((WhenClause) analyze(
+            new When(
+                new Function("=", Arrays.asList(node.getCaseValue(), when.getCondition())),
+                when.getResult()
+            ), context));
+      }
```

Comment:
Can we merge two different case...when logic in the AstBuilder?

Reply:
yeah, we can, I was thinking make AST close to original form and do this equivalent transform and type check in analyzer together.